cohttp-eio/src/server.ml

```diff
@@ -0,0 +1,260 @@
+    match tm.tm_wday with
+    | 0 -> "Sun"
+    | 1 -> "Mon"
+    | 2 -> "Tue"
+    | 3 -> "Wed"
+    | 4 -> "Thu"
+    | 5 -> "Fri"
+    | 6 -> "Sat"
+    | 7 -> "Sun"
+    | _ -> assert false
+  in
+  let month =
+    match tm.tm_mon with
+    | 0 -> "Jan"
+    | 1 -> "Feb"
+    | 2 -> "Mar"
+    | 3 -> "Apr"
+    | 4 -> "May"
+    | 5 -> "Jun"
+    | 6 -> "Jul"
+    | 7 -> "Aug"
+    | 8 -> "Sep"
+    | 9 -> "Oct"
+    | 10 -> "Nov"
+    | 11 -> "Dec"
+    | _ -> assert false
+  in
+  Format.sprintf "%s, %02d %s %04d %02d:%02d:%02d GMT" weekday tm.tm_mday month
+    (1900 + tm.tm_year) tm.tm_hour tm.tm_min tm.tm_sec
+
+(* https://datatracker.ietf.org/doc/html/rfc7230#section-4.1 *)
+let write_chunked faraday chunk_writer =
+  let write_extensions exts =
+    List.iter
+      (fun { Chunk.name; value } ->
+        Faraday.write_char faraday ';';
+        Faraday.write_string faraday name;
+        Option.iter
+          (fun v ->
+            let v = Printf.sprintf "=%s" v in
+            Faraday.write_string faraday v)
+          value)
+      exts
+  in
+  let write = function
+    | Chunk.Chunk { size; data; extensions } ->
+        let size = Printf.sprintf "%X" size in
+        Faraday.write_string faraday size;
+        write_extensions extensions;
+        Faraday.write_string faraday "\r\n";
+        Faraday.write_bigstring faraday data.buffer;
+        Faraday.write_string faraday "\r\n"
+    | Chunk.Last_chunk extensions ->
+        let size = Printf.sprintf "%X" 0 in
+        Faraday.write_string faraday size;
+        write_extensions extensions;
+        Faraday.write_string faraday "\r\n"
+  in
+  chunk_writer write
+
+let write_response (client_conn : Client_connection.t) { Response.res; body } =
+  let faraday = Faraday.create Reader.default_io_buffer_size in
+  let serialize () =
+    let status_line =
+      let version = Http.Response.version res |> Http.Version.to_string in
+      let status = Http.Response.status res |> Http.Status.to_int in
+      let status_phrase = Http.Status.reason_phrase_of_code status in
+      Printf.sprintf "%s %d %s\r\n" version status status_phrase
+    in
+    Faraday.write_string faraday status_line;
+    let headers = Http.Response.headers res in
+    let headers =
+      (*--- Don't cache set-cookie headers in browsers and proxies. ---*)
+      if Http.Header.mem headers "set-cookie" then
+        Http.Header.add headers "Cache-Control" {|no-cache="Set-Cookie"|}
+      else headers
+    in
+    let headers =
+      let hdr = "content-length" in
+      match body with
+      | `String cs ->
+          Http.Header.add_unless_exists headers hdr
+            (string_of_int @@ Cstruct.length cs)
+      | `Chunked _ -> Http.Header.remove headers hdr
+      | `Custom _ -> headers
+      | `None -> Http.Header.add_unless_exists headers hdr "0"
+    in
+    let headers =
+      let date = Unix.time () |> Unix.gmtime in
+      Http.Header.add_unless_exists headers "Date" (datetime_to_string date)
```

Comment:
Do we really need this? I don't remember `Date` being mandatory.

Reply:
According to this spec (https://www.rfc-editor.org/rfc/rfc7231#section-7.1.1.2), `Date` seems to be mandatory if the server is capable. To quote `An
   origin server MUST send a Date header field in all other cases.`

Reply:
golang http lib also seems to add this header - https://github.com/golang/go/blob/master/src/net/http/server.go#L1416-L1417

Reply:
That's a good point, but I'd still go against it. For a few reasons:

* We don't do it in the lwt/async. It's inconsistent with respect to our own libs.
* The implementation we have here is naive. Getting the current time is takes a syscall and should be cached for high performance servers.
* It's just as easy to implement outside of the basic library. It could just as easily be done through middleware. Or even a simple response generation function which would work for all backends.

Reply:
Agreed. We can drop Date for now and put it in the cohttp-consuming libraries

Reply:
Removed now.